src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/BookingTest.kt

```diff
@@ -0,0 +1,168 @@
+    bookings[4].let { it.nonArrival = nonArrivalEntityFactory.withBooking(it).produceAndPersist() }
+
+    val expectedJson = objectMapper.writeValueAsString(bookings.map(::bookingEntityToExpectedApiResponse))
+
+    val jwt = jwtAuthHelper.createValidJwt()
+
+    webTestClient.get()
+      .uri("/premises/${premises.id}/bookings")
+      .header("Authorization", "Bearer $jwt")
+      .exchange()
+      .expectStatus()
+      .isOk
+      .expectBody()
+      .json(expectedJson)
+  }
+
+  private fun bookingEntityToExpectedApiResponse(booking: BookingEntity): Booking {
+    return Booking(
+      id = booking.id,
+      person = Person(crn = booking.person!!.crn, name = booking.person!!.name),
+      arrivalDate = booking.arrivalDate,
+      departureDate = booking.departureDate,
+      keyWorker = booking.keyWorker,
+      status = entityToExpectedStatus(booking),
+      arrival = booking.arrival?.let { Arrival(bookingId = it.booking.id, arrivalDate = it.arrivalDate, expectedDepartureDate = it.expectedDepartureDate, notes = it.notes) },
+      departure = booking.departure?.let {
+        Departure(
+          bookingId = it.booking.id,
+          dateTime = it.dateTime,
+          reason = entityDepartureReasonToExpectedDepartureReason(it.reason),
+          moveOnCategory = entityMoveOnCategoryToExpectedMoveOnCategory(it.moveOnCategory),
+          destinationProvider = DestinationProvider(
+            id = it.destinationProvider.id,
+            name = it.destinationProvider.name,
+            isActive = it.destinationProvider.isActive
+          ),
+          notes = it.notes
+        )
+      },
+      nonArrival = booking.nonArrival?.let { Nonarrival(bookingId = it.booking.id, date = it.date, reason = it.reason, notes = it.notes) },
+      cancellation = booking.cancellation?.let { Cancellation(bookingId = it.booking.id, date = it.date, reason = it.reason, notes = it.notes) }
+    )
+  }
+
+  private fun entityToExpectedStatus(booking: BookingEntity): Booking.Status {
+    return if (booking.nonArrival != null) {
+      Booking.Status.notMinusArrived
+    } else if (booking.arrival != null) {
+      Booking.Status.arrived
+    } else if (booking.departure != null) {
+      Booking.Status.departed
+    } else if (booking.cancellation != null) {
+      Booking.Status.cancelled
+    } else if (booking.arrival == null && booking.nonArrival == null) {
+      Booking.Status.awaitingMinusArrival
+    } else {
+      throw RuntimeException("Could not determine status for Booking ${booking.id}")
+    }
+  }
+
+  private fun entityDepartureReasonToExpectedDepartureReason(departureReason: DepartureReason): Departure.Reason {
+    return when (departureReason) {
+      DepartureReason.Absconded -> {
+        Departure.Reason.absconded
+      }
+      DepartureReason.PlannedMove -> {
+        Departure.Reason.plannedMove
+      }
+      DepartureReason.Recalled -> {
+        Departure.Reason.recalled
+      }
+    }
+  }
```

Comment:
Is there any way we can reuse this logic from the transformers to save repeating ourselves? Or, as I think we talked about before (I think me and Ed had a difference of opinion here!) being more prescriptive about the shape of JSON we expect. We could make this easier by only returning one booking, and then testing the edge cases for transformation in unit tests for the transformer. What do you think?

Reply:
I had a similar internal debate on this one when writing - some thoughts:

```
Is there any way we can reuse this logic from the transformers to save repeating ourselves?
```

We could do this by just injecting the transformers into the test.  I'm averse to this though because we're then esentially just trusting that the transformer is correct since it will be testing its own work.  On the other hand, doing it like this we are basically just writing the same code in two different places which in practice probably isn't much better.

```
Or, as I think we talked about before (I think me and Ed had a difference of opinion here!) being more prescriptive about the shape of JSON we expect
```

I don't think writing the JSON manually provides any benefit in our case:

```
data class MyThing(
   val id: String,
   val name: String
)

val factoryProducedObject = factory<MyThing>.produceAndPersist()
val expectedJson = """
{
   "id": "${factoryProducedObject.id}",
   "name": "${factoryProducedObject.name}"
}
"""

webTestClient.get()
      .uri("/my-thing/${factoryProducedObject.id}")
      .exchange()
      .expectStatus()
      .isOk
      .expectBody()
      .json(expectedJson)
```

The `.json(expectedJson)` call doesn't just do an `expectedJson == responseBodyJson`, because we've written our JSON in a human readable way and the endpoint by default will return it without superfulous whitespace/newlines.  We would potentially also hit problems with the ordering of properties etc.  So it still needs to parse our hand-written JSON and compare it to a parsed version of the response body JSON.

With this in mind, I don't think there is an advantage in doing this over trusting Jackson's ObjectMapper (which is widely used and tested) to convert our API model (which is generated from the OpenAPI spec so we know it's the same model the endpoint is going to return.)

We could instead just get the response body as a string and compare it to our hand-written JSON.  We'd have to ensure we are ordering our hand-written JSON in the same way and not adding superfulous whitespace but I feel this would make the tests overly proscriptive  - e.g. should a Jackson (the serializer used by Spring Boot) update that alters the order in which properties are written break all of our integration tests?

We could provide all of the values as literals to the factory which would give more confidence that nothing strange is happening with us transforming betwen the generated JPA model and the API spec model:

```
data class MyThing(
   val id: String,
   val name: String
)

val factoryProducedObject = factory<MyThing>
   .withId("thing-id")
   .withName("thing-name")
   .produceAndPersist()

val expectedResponseBody = MyThing(
   id = "thing-id",
   name = "thing-name"
)

val expectedResponseBodyJson = objectMapper.writeValueAsString(expectedResponseBody)

webTestClient.get()
      .uri("/my-thing/${factoryProducedObject.id}")
      .exchange()
      .expectStatus()
      .isOk
      .expectBody()
      .json(expectedJson)
```

But this seems to invalidate the point in using KFactory to begin with - we'd be writing way more code than if we were just writing the SQL by hand but for no added benefit.


```We could make this easier by only returning one booking, and then testing the edge cases for transformation in unit tests for the transformer.```

Agreed on this, I will add some transformer unit tests so we don't need to do all the edge cases here.  I think the above discussion is still valid even if we only do one Booking in the integration test though.

Reply:
What do you think about injecting the transformers into the test, as long as we unit testing the transformers when there's any logic beyond simple marshalling?

Reply:
Fair point, seems reasonable enough if they've been unit tested :)